### DISCUSSION TITLE
Possible to test a custom event dispatched from the react component and assert on the data

### DISCUSSION THREAD

**Discussion by southbeach** [vitest-dev/vitest#6488] [Q&A]
For instance, the component will dispatch an event on load

**Component**

```js
const User = () => {
  useEffect(() => {
        dispatchEvent(new CustomEvent("mounted", {bubbles: true, detail: {foo: "bar"}}));
  }, [])
}
```

**Unit Test**

```js
it("should get event", () => {
    return new Promise(resolve => { // return promise so runner waits on it.
        window.addEventListener("mounted", event => {
            console.log("listener"); // not called
            expect(event.detail.foo).equal("bar");
            resolve();
        });

        act(() => {
            render(<User />);
        });
    });
});
```

Issue: The issue is that the callback listener is not called.

**Comment by hi-ogawa** [comment]
Can you provide a reproduction? Maybe it's possible that jsdom/happy-dom doesn't support custom even dispatch.

**Comment by southbeach** [comment]
You have a point. I'm using   "@testing-library/jest-dom": "^6.5.0", 
and I'm not sure how to confirm if the jest-dom support custom event dispatch.

**Reply by hi-ogawa** [reply]
This script is working, so the problem might be on Vitest or your tests. It would be helpful if you can provide a reproduction to investigate further.

https://stackblitz.com/edit/node-x5rpdf?file=repro.js

```ts
import jsdom from "jsdom";

const { window } = new jsdom.JSDOM(``);

window.addEventListener("mounted", (event) => {
	console.log("works!!", event, event.detail); 
});

window.dispatchEvent(
	new window.CustomEvent("mounted", { bubbles: true, detail: { foo: "bar" } }),
);
```

**Comment by southbeach** [comment]
Thanks for the example and it made my mind clear that something is wrong with my code.

It was internal issues, moving from jest to vitest, trying different things, and I noticed dispatchEvent mocked globally via vi.fn()

I restarted with a fresh project and it worked.

Thanks a lot for your help.
